kiwipy/rmq/communicator.py

```diff
@@ -262,69 +263,81 @@ async def _send_response(self, reply_to, correlation_id, response):
 
 class RmqCommunicator:
     """
-    A publisher and subscriber using aio_pika and an asyncio event loop
+    A publisher and subscriber using aio_pika and an asyncio event loop.
+
+    The communicator takes an :class:`aio_pika.Connection` as there can be multiple communicators
+    that share a single connection.
     """
 
-    def __init__(self,
-                 connection: aio_pika.Connection,
-                 message_exchange=defaults.MESSAGE_EXCHANGE,
-                 task_exchange=defaults.TASK_EXCHANGE,
-                 task_queue=defaults.TASK_QUEUE,
-                 task_prefetch_size=defaults.TASK_PREFETCH_SIZE,
-                 task_prefetch_count=defaults.TASK_PREFETCH_COUNT,
-                 queue_expires=defaults.QUEUE_EXPIRES,
-                 encoder=defaults.ENCODER,
-                 decoder=defaults.DECODER,
-                 testing_mode=False):
+    # pylint: disable=too-many-instance-attributes
+
+    _connection = None
+    _message_subscriber = None
+    _message_publisher = None
+    _default_task_queue = None  # type: Optional[tasks.RmqTaskQueue]
+
+    def __init__(
+        self,
+        # Connection parameters
+        connection_params: Union[str, dict] = None,
```

Comment:
Does it make sense to have the `connection_params` optional? Do all factories, or at least the default one, have sensible defaults that will lead to a useful connection?

Reply:
This is porting in the behaviour of `aio_pika` where all (I think there are only two) of the factories default to "amqp://guest:guest@127.0.0.1/" [1].  I like the behaviour which is why I kept it.

[1] https://aio-pika.readthedocs.io/en/latest/apidoc.html#aio_pika.connect_robust